src/LAMMPS.jl

```diff
@@ -177,240 +245,362 @@ function get_natoms(lmp::LMP)
     Int64(API.lammps_get_natoms(lmp))
 end
 
-function dtype2type(dtype::API._LMP_DATATYPE_CONST)
-    if dtype == API.LAMMPS_INT
-        type = Ptr{Int32}
-    elseif dtype == API.LAMMPS_INT_2D
-        type = Ptr{Ptr{Int32}}
-    elseif dtype == API.LAMMPS_INT64
-        type = Ptr{Int64}
-    elseif dtype == API.LAMMPS_INT64_2D
-        type = Ptr{Ptr{Int64}}
-    elseif dtype == API.LAMMPS_DOUBLE
-        type = Ptr{Float64}
-    elseif dtype == API.LAMMPS_DOUBLE_2D
-        type = Ptr{Ptr{Float64}}
-    elseif dtype == API.LAMMPS_STRING
-        type = Ptr{Cchar}
-    else
-        @assert false "Unknown dtype: $dtype"
-    end
-    return type
+function _lammps_string(ptr::Ptr)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    return Base.unsafe_string(ptr)
 end
 
-"""
-    extract_global(lmp, name, dtype=nothing)
-"""
-function extract_global(lmp::LMP, name, dtype=nothing)
-    if dtype === nothing
-        dtype = API.lammps_extract_global_datatype(lmp, name)
-    end
-    dtype = API._LMP_DATATYPE_CONST(dtype)
-    type = dtype2type(dtype)
+function _lammps_wrap(ptr::Ptr{<:Real}, shape::Integer; own=false)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    return Base.unsafe_wrap(Array, ptr, shape, own=own)
+end
 
-    ptr = API.lammps_extract_global(lmp, name)
-    ptr = reinterpret(type, ptr)
+function _lammps_wrap(ptr::Ptr{<:Ptr{T}}, shape::NTuple{2}; own=false) where T
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
 
-    if ptr !== C_NULL
-        if dtype == API.LAMMPS_STRING
-            return Base.unsafe_string(ptr)
-        end
-        # TODO: deal with non-scalar data
-        return Base.unsafe_load(ptr)
-    end
+    shape[2] == 0 && return Matrix{T}(undef, shape) # There is no data that can be wrapped
+
+    # TODO: implement a debug mode to do this assert
+    # pointers = Base.unsafe_wrap(Array, ptr, ndata)
+    # @assert all(diff(pointers) .== count*sizeof(T))
+
+    # If the pointers are evenly spaced, we can simply use the first pointer to wrap our matrix.
+    first_pointer = unsafe_load(ptr)
+    return Base.unsafe_wrap(Array, first_pointer, shape, own=own)
+
+end
+
+function _lammps_copy(ptr::Ptr{T}, shape::Integer) where T <: Real
+    ptr == C_NULL && error("Copying NULL-pointer!")
+    return [unsafe_load(ptr, i) for i in 1:shape]
```

Comment:
I don't think it's cleaner to do:

return `copy(_lammps_wrap())`

I know a bit counter to my comment earlier, but copy can lower to memcpy directly instead of hoping for the compiler to recognize and optimize this call.